libraries/ESP8266WiFi/src/WiFiClientSecureBearSSL.cpp

```diff
@@ -253,19 +253,24 @@ void WiFiClientSecureCtx::_freeSSL() {
 }
 
 bool WiFiClientSecureCtx::_clientConnected() {
-  return (_client && _client->state() == ESTABLISHED);
+  return _client && (_client->state() == ESTABLISHED);
+}
+
+bool WiFiClientSecureCtx::_engineConnected() {
+  return _clientConnected() && _handshake_done && _eng && (br_ssl_engine_current_state(_eng) != BR_SSL_CLOSED);
 }
 
 uint8_t WiFiClientSecureCtx::connected() {
-  if (available() || (_clientConnected() && _handshake_done && (br_ssl_engine_current_state(_eng) != BR_SSL_CLOSED))) {
-    return true;
+  if (!_clientConnected()) {
+    return false;
   }
-  return false;
+
+  return (_updateRecvBuffer() || _engineConnected());
```

Comment:
`_updateRecvBuffer()` is the alternate name of both `client` and `clientCtx`'s `available()`.
The logic is now
* can't write => not connected
* otherwise: same code as before: `available()` || "SSLworking"

I was thinking that the purpose of this PR was to remove `available()` from `connected()`,
It looks like now, and maybe on purpose:
- `connected()` => false when can't write on TCP socket (even if data is available for read)
- otherwise: `connected()` => `available()` or SSL engine is enabled

If the SSL engine is not working (`!_engineConnected()`), then `connected()` still returns true if data are available for read, allowing user to call `write()` which is not possible at that time (I'm still wondering if I miss something, like `_engineConnected()` can later switch to true)

Reply:
This still goes around the issue with the socket - we definitely know the connection state from the `status()`. Internals *only* check for WiFiClient::available() when reading, and state() when writing anyway, only externals are affected.

The original idea with `available()` / `_updateRecvBuf()` is to advance the state via `_run_until()`, and maybe process some other stuff. If the engine closes while inside available(), it is returning 0. In case the underlying socket closes and there's no pending data or it reaches timeout, it is also 0.

So maybe the better idea is to:
- only use internal methods with definitive results instead of ref Arduino API
- fix Arduino API back to the reference state
- add smth like lwipConnected()

?

Reply:
Yes I do mean that.

But your ideas deserve better consideration:
> - only use internal methods with definitive results instead of ref Arduino API
> - add smth like lwipConnected()

This point is delicate
> - fix Arduino API back to the reference state

because the change (`connected()` not including `available()`) fixes network errors (#4626) when a user sketch closes without reading all received data. The notions of having the right to read and the right to write must be distinct and available to user. If you consider this API change (4y ago) is still harmful regarding official Arduino API, I suggest to open a new discussion and tag it for 3.1 so it is not forgotten, as I know other users still may have comments about that (@jandrassy at least).